sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java

```diff
@@ -0,0 +1,500 @@
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+import org.apache.beam.sdk.values.PInput;
+
+import org.bson.Document;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * IO to read and write data on MongoDB.
+ * <p>
+ * <h3>Reading from MongoDB</h3>
+ * <p>
+ * <p>MongoDbIO source returns a bounded collection of String as {@code PCollection<String>}.
+ * The String is the JSON form of the MongoDB Document.</p>
+ * <p>
+ * <p>To configure the MongoDB source, you have to provide the connection URI, the database name
+ * and the collection name. The following example illustrates various options for configuring the
+ * source:</p>
+ * <p>
+ * <pre>{@code
+ *
+ * pipeline.apply(MongoDbIO.read()
+ *   .withUri("mongodb://localhost:27017")
+ *   .withDatabase("my-database")
+ *   .withCollection("my-collection")
+ *   // above three are required configuration, returns PCollection<String>
+ *
+ *   // rest of the settings are optional
+ *
+ * }</pre>
+ * <p>
+ * <p>The source also accepts an optional configuration: {@code withFilter()} allows you to
+ * define a JSON filter to get subset of data.</p>
+ * <p>
+ * <h3>Writing to MongoDB</h3>
+ * <p>
+ * <p>MongoDB sink supports writing of Document (as JSON String) in a MongoDB.</p>
+ * <p>
+ * <p>To configure a MongoDB sink, you must specify a connection {@code URI}, a {@code Database}
+ * name, a {@code Collection} name. For instance:</p>
+ * <p>
+ * <pre>{@code
+ *
+ * pipeline
+ *   .apply(...)
+ *   .apply(MongoDbIO.write()
+ *     .withUri("mongodb://localhost:27017")
+ *     .withDatabase("my-database")
+ *     .withCollection("my-collection")
+ *     .withNumSplits(30)
+ *
+ * }</pre>
+ */
+// TODO instead of JSON String, does it make sense to populate the PCollection with BSON Document or
+//  DBObject ??
+public class MongoDbIO {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MongoDbIO.class);
+
+  public static Read read() {
+    return new Read(null, null, null, null, 0);
+  }
+
+  public static Write write() {
+    return new Write(null, null, null);
+  }
+
+  private MongoDbIO() {
+  }
+
+  /**
+   * A {@link PTransform} to read data from MongoDB.
+   */
+  public static class Read extends PTransform<PBegin, PCollection<String>> {
+
+    public Read withUri(String uri) {
+      return new Read(uri, database, collection, filter, numSplits);
+    }
+
+    public Read withDatabase(String database) {
+      return new Read(uri, database, collection, filter, numSplits);
+    }
+
+    public Read withCollection(String collection) {
+      return new Read(uri, database, collection, filter, numSplits);
+    }
+
+    public Read withFilter(String filter) {
+      return new Read(uri, database, collection, filter, numSplits);
+    }
+
+    public Read withNumSplits(int numSplits) {
+      return new Read(uri, database, collection, filter, numSplits);
+    }
+
+    private final String uri;
```

Comment:
I'd like to again suggest making the Read class *not* store the variables individually. Instead do like this:
class Read {
  private final BoundedMongoDbSource source;
  public Read() { this.source = new BoundedMongoDbSource(null, null, null, null, 0); }
  public Read withUri(String uri) { return new Read(source.withUri(uri)); }
  public Read withDatabase(String database) { return new Read(source.withDatabase(database)); }
  ...
}

And move your current withBlah methods into BoundedMongoDbSource.